src/tones/tones.1
.TH tones 1 "20 Feb 1998" "Release 2" "Linux System Manual"
.SH NAME
.I tones \- a sequential tone generator program
.SH SYNOPSIS
.B tones [-8|-16] [-s samplerate] [-loop] [-w wavfile|-o file] 
.B [-i file] [-c channels] [waveform] T freq(s)|notes(s)
.SH DESCRIPTION
.I tones
generates one or more tones of various types (waveforms) and duration (T
millisecs) of the specified frequencies or notes, or mixtures of frequencies
or notes. 
Possible waveforms are, sine
(default), cosine, square, pulse, triangle, sawtooth. T is the number of
millisecs that each tone is played for. Frequencies (freq(s)) are specified
in Hertz as integers. A frequency of 0 causes T millisecs of silence to be
played. Notes are specified as the musical note letter with an
optional '#' to sharpen the note, then an octave number. Octaves run from 
C to C. Middle C is C3, the immediately preceding note is B2! The first
Octave is from C0 to C1.
.PP
Several frequencies/notes can be played at once, by specifying the 
frequencies/notes required joined by a ',' character (but no spaces!).
.PP
e.g.
.I 1000,1500,2000 
specifies that the three frequencies are played together.
.PP
Each freq specification can optionally contain a duration, by
appending ':T', where 'T' is the duration in millisecs. This duration
overrides the default duration. Also the default duration can be changed 
by using a the ':T' format on it's own - not appended to a freq spec.
.PP
e.g. 
.I 1200,600:1000
play the two freqs for 1 sec
.PP
e.g.
.I c3,e3,g3
play the C major chord
.PP
e.g. 
.I :250
set the default tone duration to 250ms
.PP
Waveforms can be specified/altered at anytime. A single waveform name
specifies that waveform to be used for all channels. Alternatively
a comma (',') seperated list of waveforms can be given to specify or alter
the waveform to use for a given channel. Ommitting a waveform in a list,
means that the previous waveform is left unchanged.
.PP
e.g. 
.I square,,triangle
specifies to use square waves for chan 1,
chan 2 is left unchanged, and triangle waves are used for channel 3.
.PP
Their are some special 'commands' that can be specified, that may be useful
in input files.
.IP echo
The rest of the line of the input file, or the rest of the command line
parameter (NB to use quotes where necessary) is output to stdout.
.IP reset
All generator points are reset to the start of the waveform buffers. This
forces subsequant generation of multiple frequencies/waveforms to be in phase.
.PP
The digital samples (either 8 or 16 bits) are played by default to the Linux
.I /dev/dsp
device at a samplerate of 22050 samples per second, in mono mode.
.PP
Fractional Hertz frequencies are not supported. Of course, only frequencies 
less than half the samplerate (number of samples/sec) can be accurately 
generated; but the program doesn't check this.
.PP
Instead of playing the output to 
.I /dev/dsp
the samples can be written to a file as raw samples (-o file) or written in
WAV format (-w wavfile). These data files can then be played back quickly
with a raw data or WAV file player (e.g. vplay) without the overhead of
actually generating the samples.
.SH OPTIONS
.IP -8\ |\ -b\ 8
set 8 bit unsigned data samples
.IP -16\ |\ -b\ 16
set 16 bit signed little-endian data samples.
.IP -a
when used in conjunction with the -o option, data is appended to the file.
.IP -c\ CHANNELS
set the maximum number of channels (concurrent played frequencies) to 
CHANNELS. The default number is 4. There is some virtue in keeping the
number of channels to a minimum.
.IP -f
when used in conjunction with the -o or -w options, any existing file is
silently overwritten.
.IP -h
display usage and help info
.IP -i\ file
read frequencies/waveforms to generate from file 'file'.
Reads from standard input if filename is '-'. Any command line specifications
are actioned before the input file is read.
.IP -loop
play the tone sequence repetitively. Forced off if writing samples to a file
with the -o or -w options.
.IP -o\ file
write out samples to a raw data file. You will have to remember the data
format, e.g. samplerate and 8/16 bit.
.IP -s\ samplerate
set the number of samples per second to samplerate. For many simple uses a
samplerate of 8000 is sufficient, making any saved data files smaller.
.IP -w\ wavfile
write samples out in WAV format to wavfile. The WAV header contains details
of whether the data is 8 or 16 bits and the sampling rate. You cannot use the
append (-a) option with WAV files.
.IP -v
be verbose
.SH EXAMPLES
.IP tones\ 50\ 1000\ 700,1200\ 800,1100,1300
generates 3 50 millisecs sine tones, the first consisting of only 1000Hz,
the second of 700Hz and 1200Hz and the third of 800Hz, 1100Hz and 1300Hz
.IP tones\ -loop\ square\ 200\ 700\ 900\ 400\ 500
generates a sequence of 4 200 millisecs square wave tones which is repeated
until the program is interupted.
.IP tones\ -w\ seq.wav\ 70\ 1016\ 1200\ 1080\ 1150\ 1016
generates a sequence of 5 70 millisecs sine tones, and instead of playing them
the samples are stored in WAV format in seq.wav which can be played by any
WAV file player.
.SH FILES
.SH
.SH SEE  ALSO
sgen, swgen, siggen, sweepgen
.SH BUGS
.SH
.SH COPYING
.I Copyright 1995, 1996, 1997, 1998 the author (see below)
.PP
The software described by this manual is covered by the GNU General
Public License, Version 2, June 1991, issued by :
.IP
Free Software Foundation, Inc.,
.br
675 Mass Ave,
.br
Cambridge, MA 02139, USA
.PP
Permission is granted to make and distribute verbatim copies of
this manual provided the copyright notice and this permission notice
are preserved on all copies.
.PP
Permission is granted to copy and distribute modified versions of this
manual under the conditions for verbatim copying, provided that the
entire resulting derived work is distributed under the terms of a
permission notice identical to this one.
.PP
Permission is granted to copy and distribute translations of this
manual into another language, under the above conditions for modified
versions, except that this permission notice may be included in
translation instead of in the original English.
.SH AUTHOR
.I Jim Jackson
.br
.I School of Computer Studies
.br
.I University of Leeds
.I The University of Leeds
.br
.I Leeds, LS2 9JT
.br
.I UK
.br
.sp
.I Email: jj@scs.leeds.ac.uk
.br


debian/changelog
multispeech (4.6.3-0bionic1) bionic; urgency=medium

  * New upstream release.

  [ Camaleón ]
  * po/es.po: Add Spanish translation of multispeech debconf template

  [ Frans Spiesschaert ]
  * po/nl.po: Add Dutch translation of multispeech debconf template

  [ Pablo Lucas Silva Santos ]
  * po/pt_BR.po: Add Brazilian Portuguese translation.

  [ Carles Pina i Estany ]
  * Fix order of adduser command in d/templates and d/po/*.
  * debian/po/ca.po: Update/add Catalan po-debconf translation.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Mon, 10 Nov 2025 02:27:46 +0300

multispeech (4.6.1-0bionic1) bionic; urgency=medium

  * New upstream release.
  * Proper section for library package.
  * Force Speech Dispatcher to reconfigure instead of stopping it.
  * Don't restart Speech Dispatcher service.
  * Nudge Speech Dispatcher only when installation succeeded.
  * Notify Speech Dispatcher when sd-multispeech module is removed.
  * DEP5 compliant debian/copyright file.
  * Debhelper compatibility version update.
  * Homepage added.
  * Conventional dh build instead of cdbs.
  * Hardening issue addressed.
  * Multi-Arch field added.
  * Separate packages for core shared library and common configuration.
  * Use "Breaks" instead of "Conflicts" in dependencies.
  * Removed wrong dependency.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Fri, 05 Jan 2024 07:31:30 +0300

multispeech (4.6.0-1bionic1) bionic; urgency=medium

  * Debian package has been separated from the upstream sources.
  * Switch to dpkg-source 3.0 format.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Wed, 15 Nov 2023 06:51:37 +0300

multispeech (4.5.2bionic1) bionic; urgency=medium

  * English ordinals support.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Tue, 14 Feb 2023 16:01:55 +0300

multispeech (4.5.1bionic1) bionic; urgency=medium

  * Hopefully increased stability.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Tue, 26 Jul 2022 19:16:19 +0300

multispeech (4.5.0bionic1) bionic; urgency=medium

  * Numbers speaking control facility.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Tue, 07 Jun 2022 19:11:10 +0300

multispeech (4.4.4bionic1) bionic; urgency=medium

  * Russian text preprocessing fix.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Sat, 05 Mar 2022 11:55:35 +0300

multispeech (4.4.3bionic1) bionic; urgency=medium

  * Fixed preprocessing of Russian text with explicit stress notation.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Sun, 20 Feb 2022 15:45:50 +0300

multispeech (4.4.2bionic1) bionic; urgency=medium

  * Ukrainian letters reading by Russian voices.
  * Russian text preprocessing minor fixes.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Sun, 17 Oct 2021 14:10:31 +0300

multispeech (4.4.1bionic1) bionic; urgency=medium

  * More sophisticated language recognition criteria.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Wed, 29 Sep 2021 16:53:18 +0300

multispeech (4.4.0bionic1) bionic; urgency=medium

  * A bit more accurate text processing.
  * Unicode normalization before passing text to a speech synthesizer.
  * Russian text preparation minor fixes.
  * Minor language detection improvement.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Sat, 25 Sep 2021 11:17:39 +0300

multispeech (4.3.0bionic1) bionic; urgency=medium

  * Enabled explicit stress notation for Russian.
  * Explicit stress notation support in ru_tts backend.
  * Russian transliteration rules enhancement.
  * Simplified configuration structure a bit.
  * Spanish detector fix.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Sat, 18 Sep 2021 20:44:54 +0300

multispeech (4.2.1bionic1) bionic; urgency=medium

  * Debian package dependency fix.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Sun, 05 Sep 2021 20:16:20 +0300

multispeech (4.2.0bionic1) bionic; urgency=medium

  * Extensive ru_tts options support.
  * Don't place comma between numbers in a text prepared for Russian TTS
    engines.
  * Set Espeak as default TTS engine for all languages.
  * Allowed comma as decimal separator for Russian TTS engines.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Fri, 03 Sep 2021 01:53:15 +0300

multispeech (4.1.2bionic1) bionic; urgency=medium

  * Real fix of startup crashes.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Mon, 23 Aug 2021 11:28:18 +0300

multispeech (4.1.1bionic1) bionic; urgency=medium

  * Fixed crash when starting in some circumstances.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Sun, 22 Aug 2021 23:42:12 +0300

multispeech (4.1.0bionic1) bionic; urgency=medium

  * Ru_tts 6.0 compatibility.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Fri, 20 Aug 2021 22:56:16 +0300

multispeech (4.0.5bionic1) bionic; urgency=medium

  * Russian text preparation minor fix.
  * Index marks support toggling facility in SD backend.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Thu, 05 Aug 2021 14:26:19 +0300

multispeech (4.0.4bionic1) bionic; urgency=medium

  * Shared library dependency clarification.
  * Runtime guessing of Speech Dispatcher version.
  * Squeeze extra blanks in speech strings.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Wed, 04 Aug 2021 16:23:11 +0300

multispeech (4.0.3bionic1) bionic; urgency=medium

  * Speech Dispatcher 0.9+ compatibility fix.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Wed, 21 Jul 2021 01:16:05 +0300

multispeech (4.0.2bionic1) bionic; urgency=low

  * Missing autoconf declaration.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Sun, 25 Oct 2020 09:34:07 +0300

multispeech (4.0.1bionic1) bionic; urgency=low

  * Package upgrade fix.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Sun, 25 Oct 2020 02:10:19 +0300

multispeech (4.0.0bionic1) bionic; urgency=low

  * Core functionality is implemented as shared library.
  * Separate executable for Speech Dispatcher backend.
  * Separate packages for Emacspeak speech server, Speech Dispatcher
    backend module, and common shared library.

 -- Igor B. Poretsky <poretsky@mlbox.ru>  Sat, 24 Oct 2020 16:35:31 +0300

multispeech (3.1.0bionic1) bionic; urgency=low

  * Conventional language designation.

src/tones/tones.c
/* tones.c
 * generate and play tones for given period....
 * Jim Jackson     Linux Version
 */

/*
 * Copyright (C) 1997 Jim Jackson                    jj@scs.leeds.ac.uk
 *                    School of Computer Studies,
 *                    The University of Leeds,
 *                    Leeds, LS2 9JT, UK
 * 
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 * 
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * 
 *  You should have received a copy of the GNU General Public License
 *  along with this program - see the file COPYING; if not, write to 
 *  the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, 
 *  MA 02139, USA.
 */

/* History...................
 * 
 * 20Feb98  Added recognition of musical notes format e.g. G3 for G
 *          in 3rd octave, A2# for A# in second octave. Octave '0'
 *          runs from C at 33Hz to C at 65Hz
 *  9Sep97  Split out the tone spec. parsing and tone playing into 
 *          tonesgen.c
 *  8Sep97  Alter tones specification to be able to specify different
 *          durations. i.e. t1,t2:ms. 
 *          Also start the change of generating technique to do sampling
 *          1Hz sample space for requisite number of samples.
 * --Aug97  Re-organise some of DAC routines out into DAC.c
 * 18Feb97  -f flag missing, added it
 *          also added facility to be able to change wavform. A
 *          non-numeric parameter will cause a change in waveform
 *          for rest of tones.
 *          Also still occasional noise after tones played - altered DSPend again!
 * 16Feb97  altered it so you can specify several freqs to be played together
 *          by using punctuation of some sort to seperate freqs 
 *          e.g. 440,880   would play two freqs of 440Hz and 880Hz together
 * 13Feb97  Based on sgen, using generator.c - only mono version.
 * 
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <ctype.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/soundcard.h>
#include <fcntl.h>
#include <math.h>


#define VERSION "tones  Ver. 1.3 (Sep 97)   tone generator"

#define DAC_FILE "/dev/dsp"
#define SAMPLERATE 22050
#define DEF_CHANNELS 4
#define EFAIL -1

#define MAXPRM 32
#define MAXTONES 64
#define chkword(a,b) ((n>=a)&&(strncmp(p,b,n)==0))

int aflg,fflg,vflg,dflg,wfmt;
int loopflg;

unsigned int stereo;             /* stereo or mono mode      */
unsigned int samplerate;         /* Samples/sec              */
unsigned int ratio;              /* used in pulse, sweep etc */
unsigned int afmt;               /* format for DSP  */
unsigned int DSPbytecount;       /* number of bytes sent to DSP  */
int frag_size;                   /* size of a DSP buffer  */
int channels;                    /* number of generating channels */

char *sys,*outfile,*infile,*dsp_device;

extern char *getGenErrStr();

help(e)
int e;
{  
   char **aa;
   
   fputs(VERSION,stdout);
   fputs("\nUsage: tones [-s samplerate] [-8/-16|-b 8/16] [waveform] T freq(s)\n",stdout);
   fprintf(stdout,"Defaults: SINE wave, %d channels, output to %s, %d samples/sec,\n",
	   DEF_CHANNELS,dsp_device,SAMPLERATE);
   fputs("          mono, 16 bit samples if possible, else 8 bit.\n\n",stdout);
   fputs("\nGenerates and plays for T millisecs, tones of given frequencies.\n",stdout);
   fputs("Several freqs can be played together by specifying a list of freqs\n",stdout);
   fputs("separated by ',' or '+', e.g. 150,200,300 or 100+200+300\n",stdout);
   fputs("The maximum number of frequencies which can be played together\n",stdout);
   fputs("is specified by the '-c channels' option. Each channel adds\n",stdout);
   fputs("processing overhead to the gneration process.\n",stdout);
   fputs("An optional duration maybe specified, that overrides the default,\n",stdout);
   fputs("by appending ':n' where n specifies number of millisecs,\n",stdout);
   fputs("e.g. 150,300:250 plays 150Hz and 300Hz together for 250 millisecs\n",stdout);
   fputs("     0:1000 plays one second (1000ms) of silence.\n",stdout);
   fputs("Valid waveforms are:",stdout);
   for ( aa=(char **)getWavNames(); *aa; aa++ ) fprintf(stdout," %s",*aa);
   fputs(".\nand can be specified anywhere and affect all later tones.\n",stdout);
   fputs("flags: -f,-a         force overwrite/append of/to file\n",stdout);
   fputs("       -o/-w file    write samples to raw/WAVE file ('-' is stdout)\n",stdout);
   fputs("       -i file       contents of file specify tones program ('-' is stdin)\n",stdout);
   fputs("       -s samples    generate with samplerate of samples/sec\n",stdout);
   fputs("       -c channels   number of channels or voices\n",stdout);
   fputs("       -loop         repeat play the tones ad nauseam.\n",stdout);
   fputs("       -v            be verbose.\n",stdout);
   fputs(" -8/-16 or -b 8|16   force 8 bit or 16 bit mode.\n",stdout);
   return(e);
}

/* main
 *
 */
 
main(argc,argv)
int argc;
char **argv;
{
   unsigned int i,j,k,l,m,n,N;
   FILE *f,*fi;
   char *p,*fnm,bf[260],*omode;
   int fd,st;
   unsigned int t;

   dsp_device = getenv("DSP_DEVICE");
   if (!dsp_device) dsp_device = DAC_FILE;
   if ((p=strrchr(sys=*argv++,'/'))!=NULL) sys=++p;
   argc--;
   
   outfile=infile=NULL;
   samplerate=SAMPLERATE; afmt=AFMT_QUERY;
   channels=DEF_CHANNELS;
   loopflg=stereo=wfmt=frag_size=0;                 /* force mono mode */
   dflg=vflg=aflg=fflg=0;
   
   while (argc && **argv=='-') {          /* all flags and options must come */
      n=strlen(p=(*argv++)+1); argc--;    /* before paramters                */
      if (chkword(1,"samplerate")) {
	 if (argc && isdigit(**argv)) { samplerate=atoi(*argv++); argc--; }
      }
      else if (chkword(1,"output")) {     /* specify output file       */
	 if (argc) { outfile=*argv++; argc--; } /* output file name          */
      }
      else if (chkword(1,"input")) {     /* specify input file       */
	 if (argc) { infile=*argv++; argc--; } /* input file name          */
      }
      else if (chkword(1,"loop")) {       /*  specify loop option   */
         loopflg=1;
      }
      else if (chkword(1,"wave")) {     /* e.g. write WAVE format file  */
	 if (argc) { outfile=*argv++; argc--; wfmt=1; } /* output file name  */
      }
      else if (chkword(2,"16")) { afmt=AFMT_S16_LE; }
      else if (chkword(1,"bits")) {
	 i=0;
	 if (argc) { 
	    i=atoi(*argv++); argc--;
	 }
	 if (i==8) afmt=AFMT_U8;
	 else if (i==16) afmt=AFMT_S16_LE;
	 else exit(err_rpt(EINVAL,"must be '-b 8' or '-b 16'."));
      }
      else if (chkword(1,"channels")) {
	 i=0;
	 if (argc) { 
	    i=atoi(*argv++); argc--;
	 }
	 if (i<1) exit(err_rpt(EINVAL,"Must have 1 or more channels."));
	 channels=i;
      }
      else {                             /* check for single char. flags    */
	 for (; *p; p++) {
	    if (*p=='h') exit(help(EFAIL));
	    else if (*p=='8') afmt=AFMT_U8;
	    else if (*p=='a') aflg=1;
	    else if (*p=='f') fflg=1;
	    else if (*p=='d') dflg=1;
	    else if (*p=='l') loopflg=1;
	    else if (*p=='v') vflg++;
	    else {
	       *bf='-'; *(bf+1)=*p; *(bf+2)=0;
	       exit(help(err_rpt(EINVAL,bf)));
	    }
	 }
      }
   }
   
   if (vflg) puts(VERSION);

   fi=NULL;            /* fi!=NULL if a command file is to be read */
   if (infile!=NULL) { /* check for valid infile if specified */
      if (strcmp(infile,"-")==0) {
	 fi=stdin; infile="stdin";
      } else if ((fi=fopen(infile,"r"))==NULL) {
	 exit(err_rpt(errno,infile));
      }
   }
   
   if (outfile==NULL) {           /* if no outfile then write direct to DAC */
                                  /* if no format specified then try 16 bit */
      i=(afmt==AFMT_QUERY)?AFMT_S16_LE:afmt ; 
      if ((fd=DACopen(fnm=dsp_device,"w",&samplerate,&i,&stereo))<0) {
	 if (afmt==AFMT_QUERY) {  /* if no format specified try for 8 bit.. */
	    i=AFMT_U8;
	    fd=DACopen(fnm=dsp_device,"w",&samplerate,&i,&stereo);
	 }
	 if (fd<0) exit(err_rpt(errno,"Opening DSP for output."));
      }
      if ((frag_size=getfragsize(fd))<0)
	exit(err_rpt(errno,"Problem getting DSP Buffer size."));

      if (vflg) {
	 printf("%s : DAC Opened for output\n",fnm);
         printf("%d bytes per DSP buffer.\n",frag_size);
      }
      if ((afmt!=AFMT_QUERY && i!=afmt) || stereo!=0) {
	 if (i!=afmt) 
	   err_rpt(EINVAL,"Sound card doesn't support requested sample format.");
	 if (stereo!=0)
	   err_rpt(EINVAL,"Card doesn't do MONO - nah, don't believe it.");
	 exit(EINVAL);
      }
      afmt=i;
   }
   else if (strcmp(outfile,"-")==0) {
      if (aflg || wfmt)
         exit(err_rpt(EINVAL,"Cannot write append, or write a WAVE file, to stdout."));
      f=stdout; fnm="stdout";
   }
   else {
      if (aflg && wfmt) {
         exit(err_rpt(EINVAL,"Cannot write append to a WAVE file."));
      }
      afmt=(afmt==AFMT_QUERY)?AFMT_S16_LE:afmt ; /* set 16 bit mode unless
						  * some format is forced */
      omode=(aflg)?"a":"w";
      if ((f=fopen(fnm=outfile,"r"))!=NULL) {
	 fclose(f);
	 if ( (!aflg) && (!fflg) ) exit(err_rpt(EEXIST,fnm));
      }
      if ((f=fopen(fnm,omode))==NULL) { exit(err_rpt(errno,fnm)); }
      frag_size=32768;
      if (vflg) { 
	 printf("%s samples to %s file %s\n",
		(*omode=='a')?"Appending":"Writing",
		(wfmt)?"WAVE":"RAW", fnm);	       
         printf("Buffer size is %d bytes.\n",frag_size);
      }
      fd=fileno(f);
      loopflg=0;    /* cannot loop if writing to a file! */
      if (wfmt && (writewavhdr(fd,0,samplerate,afmt,stereo)<0)) {
	 exit(err_rpt(errno,"Write WAVE file header."));
      }
   }

   if (!(afmt==AFMT_S16_LE || afmt==AFMT_U8)) {
      exit(err_rpt(EINVAL,"Only unsigned 8 and signed 16 bit samples supported."));
   }
 	
   if (vflg) {
      printf("%d %s, %s, samples/sec\n",samplerate,
	     (stereo)?"stereo":"mono",
	     (afmt==AFMT_U8)?"unsigned 8 bit":
	     ((afmt==AFMT_S16_LE)?"signed 16 bit, little endian":
	      "Unknown audio format"));
      printf("Upto %d Frequencies can be played together.\n",channels);
   }
   
   
   /* initialise base samples etc */
   init_samples(fd,frag_size,samplerate,afmt,stereo,channels);
   
   do {
      for ( i=0; i<argc; i++) {
	 if (vflg) printf("%s\n",argv[i]);
	 if (st=doToneCmd(argv[i])) {
	    fprintf(stderr,"%s -> %s\n",argv[i],getGenErrStr());
	 }
      }
      if (fi!=NULL) {
	 for (i=1,rewind(fi); (p=fgets(bf,260,fi))!=NULL; i++) {
	    for ( ; isspace(*p); p++) {}
	    if (*p && *p!='#') {
	       if (vflg) printf("%s\n",p);
	       if (st=doToneCmd(p)) {
		  fprintf(stderr,"[%3d] %s -> %s\n",i,p,getGenErrStr());
	       }
	    }
	 }
      }
   } while (loopflg);
   
   if (outfile==NULL) {
      flush_samples();
   } else if (wfmt && 
	      (writewavhdr(fd,getsamples(),samplerate,afmt,stereo)<0)) {
      err_rpt(errno,"Rewriting WAVE file header.");
   }
   
   close(fd);
   exit(0);
}  

doline(s)
char *s;

{
   int i,n;
   char **aa[50];
   
   for ( ; isspace(*s); s++) {}
   if (*s && *s!='#') {
      n=parse(s,aa,'\0');
      for (i=0; i<n; i++) doToneCmd(aa[i]);
   }
}
